src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -64,146 +76,259 @@ makeSem = genFreer True
+  } deriving Show
+
+------------------------------------------------------------------------------
+-- | Creates info about smart constructor being created from info about action
+-- and it's parent type.
+mkCLInfo :: DatatypeInfo -> ConstructorInfo -> Q ConLiftInfo
+mkCLInfo dti ci = do
+  let cliEffName            = datatypeName dti
+
+  (raw_cli_eff_args, [m_arg, raw_cli_res_arg]) <-
+    case splitFromEnd 2 $ datatypeInstTypes dti of
+      r@(_, [_, _]) -> return r
+      _             -> missingEffTArgs cliEffName
+
+  m_name <-
+    case tVarName m_arg of
+      Just r        -> return r
+      Nothing       -> mArgNotVar cliEffName m_arg
+
+  cliUnionName <- newName "r"
+
+  let normalizeType         = replaceMArg m_name cliUnionName
+                            . simplifyKinds
+                            . applySubstitution eq_pairs
+      -- We extract equality constraints with variables to unify them
+      -- manually - this makes type errors more readable. Plus we replace
+      -- kind of result with 'Type' if it is a type variable.
+      (eq_pairs, cliFunCxt) = first (M.fromList . maybeResKindToType)
+                            $ partitionEithers
+                            $ eqPairOrCxt <$> constructorContext ci
+      -- TODO: take result type by name or position?
+      maybeResKindToType    = maybe id (\k ps -> (k, StarT) : ps)
+                            $ tVarName $ tvKind $ last
+                            $ datatypeVars dti
+
+      cliEffArgs            = normalizeType <$> raw_cli_eff_args
+      cliResType            = normalizeType     raw_cli_res_arg
+      cliConName            = constructorName ci
+      cliFunName            = mapNameBase (mapHead toLower)
+                            $ constructorName ci
+      cliFunArgs            = normalizeType <$> constructorFields ci
+
+  return CLInfo{..}
+
+------------------------------------------------------------------------------
+-- Error messages and checks
+
+mArgNotVar :: Name -> Type -> Q a
+mArgNotVar name mArg = fail $ show
+  $  text "Monad argument ‘" <> ppr mArg <> text "’ in effect ‘"
+  <> ppr name <> text "’ is not a type variable"
+
+missingEffTArgs :: Name -> Q a
+missingEffTArgs name = fail $ show
+  $   text "Effect ‘" <> ppr name
+      <> text "’ has not enough type arguments"
+  $+$ nest 4
+      (   text "At least monad and result argument are required, e.g.:"
+      $+$ nest 4
+          (   text ""
+          $+$ ppr (DataD [] base args Nothing [] []) <+> text "..."
+          $+$ text ""
+          )
+      )
+  where
+    base = mapNameBase id name
+    args = PlainTV . mkName <$> ["m", "a"]
+
+checkExtensions :: [Extension] -> Q ()
+checkExtensions exts = do
+  states <- zip exts <$> traverse isExtEnabled exts
+  maybe (return ())
+        (\(ext, _) -> fail $ show
+          $ char '‘' <> text (show ext) <> char '’'
+            <+> text "extension needs to be enabled\
```

Comment:
what is this trailing sorcery?

Reply:
79-character limit :smile: Should I format it differently?

Reply:
I didn't know about this feature, so I'm mostly surprised. I don't love it, but maybe I will come to, in time.